File: go/leet_code/2441.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

import "slices"

// findMaxK finds the maximum integer 'k' in the given slice 'nums' such that both 'k' and '-k' exist in 'nums'.
// If no such integer exists, it returns -1.
//
// It first sorts the slice and then checks for each number from the end if its negative counterpart exists in the slice.
//
// Parameters:
// nums: A slice of integers. It is expected to be non-empty.
//
// Returns:
// The maximum integer 'k' such that both 'k' and '-k' exist in 'nums'. If no such integer exists, it returns -1.
func findMaxK(nums []int) int {
	// Sort the slice in ascending order.
	slices.Sort(nums)

	// Create a map to store the existence of each number in the slice.
	m := make(map[int]bool, len(nums))

	// Populate the map with the numbers in the slice.
	for _, num := range nums {
		m[num] = true
	}

	// Check for each number from the end if its negative counterpart exists in the slice.
	for i := len(nums) - 1; i > 0; i-- {
		if m[-nums[i]] {
			// If the negative counterpart exists, return the number.
			return nums[i]
		}
	}
	// If no such number exists, return -1.
	return -1
}

File: typescript/leet_code/2441.ts
/**
 * @file This module contains a function to find the maximum absolute value in an array that also has its negative counterpart.
 * @module leet_code/2441
 */

/**
 * Finds the maximum absolute value in an array that also has its negative counterpart.
 *
 * This function iterates over the array of numbers, maintaining a set of previously encountered numbers.
 * For each number, it checks if the negative of its absolute value (for positive numbers) or its absolute value (for negative numbers)
 * is in the set of previously encountered numbers. If it is, it updates the maximum value found so far.
 *
 * @param {number[]} nums - The array of numbers to search. It is expected to contain both positive and negative numbers.
 * @returns {number} The maximum absolute value in the array that also has its negative counterpart, or -1 if no such number exists.
 *
 * @throws {TypeError} If the input is not an array of numbers.
 */
function findMaxK(nums: number[]): number {
    if (!Array.isArray(nums) || !nums.every(num => typeof num === 'number')) {
        throw new TypeError('Input must be an array of numbers');
    }

    let max: number = -1;
    const previousNums: Set<number> = new Set<number>();

    for (const num of nums) {
        if (num > 0) {
            const negative: number = -Math.abs(num);
            // If the negative counterpart of the current number exists in the set, update the max value
            if (previousNums.has(negative)) {
                max = Math.max(max, num);
            }
        } else {
            const positive: number = Math.abs(num);
            // If the positive counterpart of the current number exists in the set, update the max value
            if (previousNums.has(positive)) {
                max = Math.max(max, positive);
            }
        }
        // Add the current number to the set of previously encountered numbers
        previousNums.add(num);
    }

    return max;
}
